x/bank/keeper/keeper.go

```diff
@@ -619,6 +623,14 @@ func (k BaseKeeper) destroyCoins(ctx sdk.Context, moduleName string, amounts sdk
 // It will panic if the module account does not exist or is unauthorized.
 func (k BaseKeeper) BurnCoins(ctx sdk.Context, moduleName string, amounts sdk.Coins) error {
 	subFn := func(ctx sdk.Context, moduleName string, amounts sdk.Coins) error {
+
+		// Try subtract from the in mem var first, prevents the condition where
+		// the module may have a pending deposit that would be enough to pay for this send
+		ok := ctx.ContextMemCache().SafeSubDeferredSends(moduleName, amounts)
```

Comment:
is it safe to add this in the normal BurnCoins and not the DeferredBurn? Doesn't that make the deferred mem resources a potential race condition?

Reply:
This is in the case where we call deferredMint and then call the regular burn, in this case if we don't check the deferred sends then it will error out with insufficient balance. 

The deferred in mem variables are all blocked by mutex, I think in this case we only care about the absolute value of the in-mem variables at the end of FinalizeBlocker, as only a single value is written to the store

Reply:
gotcha, makes sense